TypeNotPresentExceptionProxy

### TestNG Version

6.10

### Expected behavior

Tests complete as expected in Jenkins

### Actual behavior

Tests run fine locally, but with Jenkins after the tests are run a `TypeNotPresentExceptionProxy` is thrown for some reason.

### Is the issue reproductible on runner?

- [ ] Shell
- [X] Maven on Jenkins
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample

No test case. But here's a Jenkins run where it happens. https://builds.apache.org/job/Curator/63/console

Here's the stacktrace:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project curator-test-zk34: ExecutionException There was an error in the forked process
[ERROR] java.lang.ArrayStoreException: sun.reflect.annotation.TypeNotPresentExceptionProxy
[ERROR] at sun.reflect.annotation.AnnotationParser.parseClassArray(AnnotationParser.java:724)
[ERROR] at sun.reflect.annotation.AnnotationParser.parseArray(AnnotationParser.java:531)
[ERROR] at sun.reflect.annotation.AnnotationParser.parseMemberValue(AnnotationParser.java:355)
[ERROR] at sun.reflect.annotation.AnnotationParser.parseAnnotation2(AnnotationParser.java:286)
[ERROR] at sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:120)
[ERROR] at sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:72)
[ERROR] at java.lang.reflect.Executable.declaredAnnotations(Executable.java:599)
[ERROR] at java.lang.reflect.Executable.declaredAnnotations(Executable.java:597)
[ERROR] at java.lang.reflect.Executable.getAnnotation(Executable.java:570)
[ERROR] at java.lang.reflect.Method.getAnnotation(Method.java:622)
[ERROR] at org.testng.internal.annotations.JDK15AnnotationFinder.findAnnotation(JDK15AnnotationFinder.java:111)
[ERROR] at org.testng.internal.TestNGClassFinder.<init>(TestNGClassFinder.java:77)
[ERROR] at org.testng.internal.TestNGClassFinder.<init>(TestNGClassFinder.java:40)
[ERROR] at org.testng.TestRunner.initMethods(TestRunner.java:403)
[ERROR] at org.testng.TestRunner.init(TestRunner.java:252)
[ERROR] at org.testng.TestRunner.init(TestRunner.java:222)
[ERROR] at org.testng.TestRunner.<init>(TestRunner.java:163)
[ERROR] at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:585)
[ERROR] at org.testng.SuiteRunner.init(SuiteRunner.java:189)
[ERROR] at org.testng.SuiteRunner.<init>(SuiteRunner.java:136)
[ERROR] at org.testng.TestNG.createSuiteRunner(TestNG.java:1375)
```

I posted this issue at Surefire (https://issues.apache.org/jira/browse/SUREFIRE-1612) and they insist the problem is in TestNg. I also just tried with the latest beta `7.0.0-beta1` and it fails in the same manner. So, I'm re-opening this. Here's is the full stacktrace from a run this morning:

```
[ERROR] ExecutionException There was an error in the forked process
[ERROR] sun.reflect.annotation.TypeNotPresentExceptionProxy
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: ExecutionException There was an error in the forked process
[ERROR] sun.reflect.annotation.TypeNotPresentExceptionProxy
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.awaitResultsDone(ForkStarter.java:494)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.runSuitesForkPerTestSet(ForkStarter.java:441)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:293)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:245)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1194)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1022)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:868)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:956)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:290)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:194)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR] Caused by: org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
[ERROR] sun.reflect.annotation.TypeNotPresentExceptionProxy
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:658)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:533)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.access$600(ForkStarter.java:115)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter$2.call(ForkStarter.java:429)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter$2.call(ForkStarter.java:406)
[ERROR] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[ERROR] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[ERROR] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[ERROR] 	at java.lang.Thread.run(Thread.java:745)
[ERROR] 
```

@Tibor17 - The error is stemming from the Java provided reflection APIs that TestNG is using to extract out various different information.
As @juherr also pointed out, this is usually due to a messed up class path.

@Randgalt - I ran the entire codebase (it did take a lot of time to run to completion) but I was able to simulate the problem on my local machine as well. So this is NOT a Jenkins specific problem for sure.

```
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:37 h
[INFO] Finished at: 2018-12-18T01:15:21+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (default-test) on project curator-test-zk34: There are test failures.
[ERROR]
[ERROR] Please refer to /Users/krmahadevan/temp/curator/curator-test-zk34/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] ExecutionException There was an error in the forked process
[ERROR] sun.reflect.annotation.TypeNotPresentExceptionProxy
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: ExecutionException There was an error in the forked process
[ERROR] sun.reflect.annotation.TypeNotPresentExceptionProxy
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.awaitResultsDone(ForkStarter.java:494)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.runSuitesForkPerTestSet(ForkStarter.java:441)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:293)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:245)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1194)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1022)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:868)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:956)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:192)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR] Caused by: org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
[ERROR] sun.reflect.annotation.TypeNotPresentExceptionProxy
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:658)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:533)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.access$600(ForkStarter.java:115)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter$2.call(ForkStarter.java:429)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter$2.call(ForkStarter.java:406)
[ERROR] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[ERROR] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[ERROR] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[ERROR] 	at java.lang.Thread.run(Thread.java:748)
[ERROR]
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :curator-test-zk34
✘-1 ~/temp/curator [master|✔]
```

To reproduce the problem, try cleaning up the relevant folders from your local `~/.m2/repository`. It should happen for you as well.

As you guessed as well, I am pretty sure that the `curator-test-zk34` is where you need to look for to understand where is this coming from (Since you perhaps have a better idea of the codebase than me).

I dont think TestNG has anything to do here in terms of fixing this issue.

I also don't think that surefire plugin maybe at fault here, because someone had faced this problem when using ant also. Please see [this](http://vshynkliar.blogspot.com/2014/09/sunreflectannotationtypenotpresentexcep.html) blog.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-1976: Add a proper message when types are missing during annotation parsing (Krishnan Mahadevan)
 Fixed: GITHUB-2000: ThreadPoolExecutor ConcurrentModificationException (Krishnan Mahadevan)
 Fixed: GITHUB-2022: Suite.xml files make web request when suite uses DTD over HTTPS (Krishnan Mahadevan)
 Fixed: GITHUB-2017: Don't use child injectors for Guice tests (Joe Barnett)

--- a/src/main/java/org/testng/internal/annotations/AnnotationHelper.java
+++ b/src/main/java/org/testng/internal/annotations/AnnotationHelper.java
@@ -8,6 +8,7 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 
+import java.util.Objects;
 import org.testng.ITestNGMethod;
 import org.testng.annotations.IAnnotation;
 import org.testng.annotations.IConfigurationAnnotation;
@@ -29,6 +30,7 @@ import org.testng.xml.XmlTest;
  */
 public class AnnotationHelper {
   private static final Logger LOGGER = Logger.getLogger(AnnotationHelper.class);
+  private static final String SUFFIX = "Please check if all classes being referred to, in the annotation are available in the classpath.";
 
   private static final List<Class<? extends IAnnotation>> ALL_ANNOTATIONS =
       Arrays.asList(
@@ -345,4 +347,41 @@ public class AnnotationHelper {
 
     return result.toString();
   }
-}
+
+  static <A extends Annotation> A getAnnotationFromClass(Class<?> clazz, Class<A> a) {
+    Objects.requireNonNull(clazz, "Cannot retrieve annotations from a null class.");
+    Objects.requireNonNull(a, "Cannot work with a null annotation");
+    try {
+      return clazz.getAnnotation(a);
+    } catch (Throwable t) {
+      String msg = "Encountered problems when parsing the annotation on class " + clazz.getCanonicalName() + ". ";
+      msg += SUFFIX;
+      throw new TypeNotPresentException(msg, t);
+    }
+  }
+
+  static <A extends Annotation> A getAnnotationFromMethod(Method method, Class<A> a) {
+    Objects.requireNonNull(method, "Cannot retrieve annotations from a null class.");
+    Objects.requireNonNull(a, "Cannot work with a null annotation");
+    try {
+      return method.getAnnotation(a);
+    } catch (Throwable t) {
+      String mname = method.getDeclaringClass().getCanonicalName() + "." + method.getName();
+      String msg = "Encountered problems when parsing the annotation on method " + mname + "(). ";
+      msg += SUFFIX;
+      throw new TypeNotPresentException(msg, t);
+    }
+  }
+
+  static <A extends Annotation> A getAnnotationFromConstructor(Constructor<?> c, Class<A> a) {
+    Objects.requireNonNull(c, "Cannot retrieve annotations from a null class.");
+    Objects.requireNonNull(a, "Cannot work with a null annotation");
+    try {
+      return c.getAnnotation(a);
+    } catch (Throwable t) {
+      String msg = "Encountered problems when parsing the annotation on constructor " + c.getName() + ". ";
+      msg += SUFFIX;
+      throw new TypeNotPresentException(msg, t);
+    }
+  }
+ }

--- a/src/main/java/org/testng/internal/annotations/JDK15AnnotationFinder.java
+++ b/src/main/java/org/testng/internal/annotations/JDK15AnnotationFinder.java
@@ -77,10 +77,12 @@ public class JDK15AnnotationFinder implements IAnnotationFinder {
     // because inheritance of this annotation causes aggregation instead of
     // overriding
     if (a.equals(org.testng.annotations.Listeners.class)) {
-      return cls.getAnnotation(a);
+
+      return AnnotationHelper.getAnnotationFromClass(cls, a);
     } else {
       while (cls != null) {
-        A result = cls.getAnnotation(a);
+        A result = AnnotationHelper.getAnnotationFromClass(cls, a);
+
         if (result != null) {
           return result;
         } else {
@@ -105,7 +107,7 @@ public class JDK15AnnotationFinder implements IAnnotationFinder {
       throw new IllegalArgumentException(
           "Java @Annotation class for '" + annotationClass + "' not found.");
     }
-    Annotation annotation = m.getAnnotation(a);
+    Annotation annotation = AnnotationHelper.getAnnotationFromMethod(m, a);
     return findAnnotation(
         m.getDeclaringClass(),
         annotation,
@@ -129,9 +131,9 @@ public class JDK15AnnotationFinder implements IAnnotationFinder {
     if (tm.getInstance() != null) {
       testClass = tm.getInstance().getClass();
     }
-    Annotation annotation = m.getAnnotation(a);
+    Annotation annotation = AnnotationHelper.getAnnotationFromMethod(m, a);
     if (annotation == null) {
-      annotation = testClass.getAnnotation(a);
+      annotation = AnnotationHelper.getAnnotationFromClass(testClass, a);
     }
     return findAnnotation(
         testClass, annotation, annotationClass, null, null, m, new Pair<>(annotation, m), null);
@@ -229,7 +231,7 @@ public class JDK15AnnotationFinder implements IAnnotationFinder {
       throw new IllegalArgumentException(
           "Java @Annotation class for '" + annotationClass + "' not found.");
     }
-    Annotation annotation = cons.getAnnotation(a);
+    Annotation annotation = AnnotationHelper.getAnnotationFromConstructor(cons, a);
     return findAnnotation(
         cons.getDeclaringClass(),
         annotation,

--- a/src/main/java/org/testng/internal/annotations/JDK15TagFactory.java
+++ b/src/main/java/org/testng/internal/annotations/JDK15TagFactory.java
@@ -519,7 +519,7 @@ public class JDK15TagFactory {
 
     // Not found, look on the class and then up the hierarchy
     while (cls != null && cls != Object.class) {
-      Annotation annotation = cls.getAnnotation(annotationClass);
+      Annotation annotation = AnnotationHelper.getAnnotationFromClass(cls, annotationClass);
       if (annotation != null) {
         T result = (T) invokeMethod(annotation, methodName);
         if (!def.isDefault(result)) {
@@ -546,7 +546,7 @@ public class JDK15TagFactory {
     List<String> result = Lists.newArrayList();
 
     while (cls != null && cls != Object.class) {
-      Annotation annotation = cls.getAnnotation(Test.class);
+      Annotation annotation = AnnotationHelper.getAnnotationFromClass(cls, Test.class);
       if (annotation != null) {
         String[] g = (String[]) invokeMethod(annotation, methodName);
         result.addAll(Arrays.asList(g));

--- a/src/test/java/test/JUnitTestClassLoader.java
+++ b/src/test/java/test/JUnitTestClassLoader.java
@@ -1,27 +1,16 @@
 package test;
 
 import java.io.File;
-import java.io.IOException;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.StandardOpenOption;
-import javax.tools.JavaCompiler;
-import javax.tools.ToolProvider;
-
+import java.util.List;
 import org.testng.*;
 import org.testng.annotations.Test;
+import testhelper.CompiledCode;
+import testhelper.SimpleCompiler;
+import testhelper.SourceCode;
+
 import static org.testng.Assert.*;
 
 public class JUnitTestClassLoader extends ClassLoader {
-
-  private static Path createTempDir() {
-    try {
-      return Files.createTempDirectory("junitclassloader");
-    } catch (IOException e) {
-      throw new TestNGException(e);
-    }
-
-  }
 
   @Test
   public void testPassAndFail() throws Exception {
@@ -93,17 +82,10 @@ public class JUnitTestClassLoader extends ClassLoader {
   }
 
   private Class<?> compile(String src, String name) throws Exception {
-    // compile class and load it into by a custom classloader
-    File directory = createTempDir().toFile();
-    File srcFile = new File(directory, name + ".java");
-    Files.write(srcFile.toPath(), src.getBytes(), StandardOpenOption.CREATE_NEW);
-    JavaCompiler javac = ToolProvider.getSystemJavaCompiler();
-    assertEquals(0, javac.run(null, null, null, srcFile.getCanonicalPath()));
-    srcFile.delete();
-    File classFile = new File(directory, name + ".class");
-
-    byte[] bytes = Files.readAllBytes(classFile.toPath());
-    classFile.delete();
+    File directory = SimpleCompiler.createTempDir();
+    SourceCode sourceCode = new SourceCode(name,src, directory, false);
+    List<CompiledCode> compiledCode = SimpleCompiler.compileSourceCode(sourceCode);
+    byte[] bytes = compiledCode.get(0).getByteCode();
     return defineClass(name, bytes, 0, bytes.length);
   }
 

--- a/src/test/java/test/mannotation/issue1976/IssueTest.java
+++ b/src/test/java/test/mannotation/issue1976/IssueTest.java
@@ -1,0 +1,132 @@
+package test.mannotation.issue1976;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import org.testng.TestNG;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.testng.collections.Lists;
+import testhelper.CompiledCode;
+import testhelper.SimpleCompiler;
+import testhelper.SourceCode;
+
+public class IssueTest extends ClassLoader {
+
+  private final File dir = SimpleCompiler.createTempDir();
+
+  @Test(dataProvider = "dp", expectedExceptions = TypeNotPresentException.class,
+      description = "GITHUB-1976")
+  public void testMethod(SourceCode... sources) throws IOException, ClassNotFoundException {
+    TestNG tng = new TestNG(false);
+    List<CompiledCode> byteCodes = SimpleCompiler.compileSourceCode(sources);
+    List<Class<?>> classes = Lists.newArrayList();
+    for (CompiledCode byteCode : byteCodes) {
+      if (byteCode.isSkipLoading()) {
+        continue;
+      }
+      Class<?> clazz = defineClass(byteCode.getName(), byteCode.getByteCode(), 0,
+          byteCode.getByteCode().length);
+      tng.addClassLoader(clazz.getClassLoader());
+      clazz.getClassLoader().loadClass(clazz.getName());
+      classes.add(clazz);
+    }
+    tng.setTestClasses(classes.toArray(new Class[0]));
+    tng.run();
+    Arrays.stream(sources).forEach(sourceCode -> sourceCode.getLocation().delete());
+  }
+
+  @DataProvider(name = "dp")
+  public Object[][] getTestData() throws IOException {
+    return new Object[][]{
+        {new SourceCode[]{missingTypeAtMethodLevel(), exception1()}},
+        {new SourceCode[]{missingTypeAtClassLevel(), exception2()}},
+        {new SourceCode[]{missingTypeAtConstructor(), dataProvider()}},
+        {new SourceCode[]{missingTypeAtListenerAnnotation(), listener()}},
+        {new SourceCode[]{missingTypeAtBaseClass(), childClass(), exception3()}},
+    };
+  }
+
+  private SourceCode missingTypeAtMethodLevel() throws IOException {
+    String source = "import org.testng.annotations.Test\n;";
+    source += "public class Sample1 {\n";
+    source += "  @Test(expectedExceptions = MyFancyException.class)\n";
+    source += "  public void testMethod() {\n";
+    source += "  }\n";
+    source += "}\n";
+    return new SourceCode("Sample1", source, dir, false);
+  }
+
+  private SourceCode exception1() throws IOException {
+    String source = "public class MyFancyException extends RuntimeException {}";
+    return new SourceCode("MyFancyException", source, dir, true);
+  }
+
+  private SourceCode missingTypeAtClassLevel() throws IOException {
+    String source = "import org.testng.annotations.Test\n;";
+    source += "@Test(expectedExceptions = AnotherException.class)\n";
+    source += "public class Sample2 {\n";
+    source += "  public void testMethod() {\n";
+    source += "  }\n";
+    source += "}\n";
+    return new SourceCode("Sample2", source, dir, false);
+  }
+
+  private SourceCode missingTypeAtBaseClass() throws IOException {
+    String source = "@org.testng.annotations.Test(expectedExceptions = ThirdException.class)\n";
+    source += "public class MyBaseClass {}\n";
+    return new SourceCode("MyBaseClass", source, dir, false);
+  }
+
+  private SourceCode childClass() throws IOException {
+    String source = "public class ChildClass extends MyBaseClass {\n;";
+    source += "public void testMethod() {}\n";
+    source += "}\n";
+    return new SourceCode("ChildClass", source, dir, false);
+  }
+
+
+  private SourceCode exception2() throws IOException {
+    String source = "public class AnotherException extends RuntimeException {}";
+    return new SourceCode("AnotherException", source, dir, true);
+  }
+
+  private SourceCode exception3() throws IOException {
+    String source = "public class ThirdException extends RuntimeException {}";
+    return new SourceCode("ThirdException", source, dir, true);
+  }
+
+  private SourceCode missingTypeAtConstructor() throws IOException {
+    String source = "import org.testng.annotations.Factory;\n";
+    source += "public class FactorySample {\n";
+    source += "  @Factory(dataProviderClass = ExampleDP.class)\n";
+    source += "  public FactorySample(Object object) { }\n";
+    source += "}\n";
+    return new SourceCode("FactorySample", source, dir, false);
+  }
+
+  private SourceCode dataProvider() throws IOException {
+    String source = "import org.testng.annotations.DataProvider;\n";
+    source += "public class ExampleDP {\n";
+    source += "  @DataProvider\n";
+    source += "  public Object[][] dp() {\n";
+    source += "    return new Object[][] {{}};";
+    source += "  }\n";
+    source += "}";
+    return new SourceCode("ExampleDP", source, dir, true);
+  }
+
+  private SourceCode missingTypeAtListenerAnnotation() throws IOException {
+    String source = "@org.testng.annotations.Listeners(SampleListener.class)\n";
+    source += "public class Sample4 {}\n";
+    return new SourceCode("Sample4", source, dir, false);
+  }
+
+  private SourceCode listener() throws IOException {
+    String source = "public class SampleListener implements org.testng.ITestListener {}";
+    return new SourceCode("SampleListener", source, dir, true);
+  }
+
+
+}

--- a/src/test/java/testhelper/CompiledCode.java
+++ b/src/test/java/testhelper/CompiledCode.java
@@ -1,0 +1,31 @@
+package testhelper;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import javax.tools.JavaFileObject.Kind;
+
+public class CompiledCode {
+  private final byte[] byteCode;
+  private final boolean skipLoading;
+  private final String name;
+
+  public CompiledCode(String name, File directory, boolean skipLoading) throws IOException {
+    this.skipLoading = skipLoading;
+    this.name = name;
+    File classFile = new File(directory, name + Kind.CLASS.extension);
+    this.byteCode = Files.readAllBytes(classFile.toPath());
+  }
+
+  public byte[] getByteCode() {
+    return byteCode;
+  }
+
+  public boolean isSkipLoading() {
+    return skipLoading;
+  }
+
+  public String getName() {
+    return name;
+  }
+}

--- a/src/test/java/testhelper/SimpleCompiler.java
+++ b/src/test/java/testhelper/SimpleCompiler.java
@@ -1,0 +1,44 @@
+package testhelper;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.List;
+import javax.tools.JavaCompiler;
+import javax.tools.ToolProvider;
+import org.testng.TestNGException;
+import org.testng.collections.Lists;
+
+public class SimpleCompiler {
+
+  public static List<CompiledCode> compileSourceCode(SourceCode... sources) throws IOException {
+    JavaCompiler javac = ToolProvider.getSystemJavaCompiler();
+    String[] arguments = new String[sources.length];
+    int i = 0;
+    for (SourceCode source : sources) {
+      arguments[i++] = source.getLocation().getCanonicalPath();
+    }
+    int rc = javac.run(null, null, null, arguments);
+    if (rc != 0) {
+      throw new TestNGException("Encountered errors when compiling code");
+    }
+    List<CompiledCode> compiledCodes = Lists.newArrayList();
+    for (SourceCode source : sources) {
+      source.getLocation().delete();
+      CompiledCode compiledCode = new CompiledCode(source.getName(), source.getDirectory(),
+          source.isSkipLoading());
+      compiledCodes.add(compiledCode);
+    }
+    return compiledCodes;
+  }
+
+  public static File createTempDir() {
+    try {
+      return Files.createTempDirectory("custom").toFile();
+    } catch (IOException e) {
+      throw new TestNGException(e);
+    }
+
+  }
+
+}

--- a/src/test/java/testhelper/SourceCode.java
+++ b/src/test/java/testhelper/SourceCode.java
@@ -1,0 +1,43 @@
+package testhelper;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import javax.tools.JavaFileObject.Kind;
+
+public class SourceCode {
+
+  private final String name;
+  private final File directory;
+  private final boolean skipLoading;
+  private final File location;
+
+  public SourceCode(String name, String src, File directory, boolean skipLoading)
+      throws IOException {
+    this.name = name;
+    this.directory = directory;
+    this.skipLoading = skipLoading;
+    this.location = new File(directory, name + Kind.SOURCE.extension);
+    if (this.location.exists()) {
+      this.location.delete();
+    }
+    Files.write(location.toPath(), src.getBytes(), StandardOpenOption.CREATE_NEW);
+  }
+
+  public File getDirectory() {
+    return directory;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public boolean isSkipLoading() {
+    return skipLoading;
+  }
+
+  public File getLocation() {
+    return location;
+  }
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -95,6 +95,7 @@
       <class name="test.failures.FailuresTest" />
       <class name="test.mannotation.MAnnotationSampleTest" />
       <class name="test.mannotation.MAnnotation2SampleTest" />
+      <class name="test.mannotation.issue1976.IssueTest" />
       <class name="test.thread.FactoryTest" />
       <class name="test.thread.SequentialTest" />
       <class name="test.thread.ParallelTestTest" />

